pip/index.py

```diff
@@ -486,9 +486,32 @@ def find_requirement(self, req, upgrade):
             c for c in all_candidates if str(c.version) in compatible_versions
         ]
 
+        yanked_candidates = [
+            c for c in applicable_candidates
+            if getattr(c.location, "yanked", False)
+        ]
+
         if applicable_candidates:
             best_candidate = max(applicable_candidates,
                                  key=self._candidate_sort_key)
+            # If we cannot find a non-yanked candidate,
+            # use the best one and print a warning about it.
+            # Otherwise, try to find another best candidate, ignoring
+            # all the yanked releases.
+            if best_candidate in yanked_candidates:
+                if set(applicable_candidates) <= set(yanked_candidates):
+                    logger.warning("WARNING: The candidate selected for download or install "
+                                   "is a yanked version: '%s' candidate (version %s at %s)\n"
+                                   "Reason for being yanked: %s",
```

Comment:
The reason may be empty; in that case this line shouldn't be there at all.

Reply:
Fixed in fixup commit. Tested on pytracemalloc:
```
$ pip install "pytracemalloc==1.4"
Collecting pytracemalloc==1.4
  WARNING: The candidate selected for download or install is a yanked version: 'pytracemalloc' candidate (version 1.4 at https://files.pythonhosted.org/packages/50/c0/e2a130c0ffdabb010577981201e98580dd1905ae1d546c02f0bef2a99087/pytracemalloc-1.4.tar.gz#sha256=c54b630c413fafb2b4605c7d76c4b54afb1f4b2e58fd16bc4365558a869111ce (from https://pypi.org/simple/pytracemalloc/))
  Using cached https://files.pythonhosted.org/packages/50/c0/e2a130c0ffdabb010577981201e98580dd1905ae1d546c02f0bef2a99087/pytracemalloc-1.4.tar.gz
    Complete output from command python setup.py egg_info:
    tracemalloc is now part of Python 3.4!
    Third party pytracemalloc module is no more needed.
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-qisocyuc/pytracemalloc/
You are using pip version 9.0.3, however version 21.2.4 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```